### PR TITLE
Deprecate alternative JVM plugins

### PR DESCRIPTION
**Pull request by jjohannes**
See #9837 

Notes: 

- I added `@Deprecated` to all API types that are exclusively used by the deprecated plugins. In particular, I checked everything that was still incubating in the corresponding projects, because a type that has been around since Gradle 2.x/3.x should either be deprecated or be de-incubated. There is one type, `JavadocArtifact`, which I left incubating but did not deprecate (yet) because it is also used in the artifact query API (which we will probably also deprecate soonish).
- I did not add any docs here. I think mentioning this in the release notes / upgrade guide is only confusing, since these should have been deprecated for a long time. I don't want users to get confuesed. The names are all so similar: _java_, _java-library_, _java-lang_, ...

### CODE REVIEW

File: subprojects/language-scala/src/main/java/org/gradle/language/scala/tasks/BaseScalaCompileOptions.java [hunk 1]
```diff
@@ -31,7 +30,6 @@
 /**
  * Options for Scala platform compilation.
  */
-@Incubating
```
Comment (jjohannes) [inline subprojects/language-scala/src/main/java/org/gradle/language/scala/tasks/BaseScalaCompileOptions.java]: This type is also used by the "normal" scala plugin and is thus not deprecated. It should not be incubating.

File: subprojects/language-java/src/main/java/org/gradle/language/java/plugins/JavaLanguagePlugin.java [hunk 2]
```diff
@@ -77,10 +78,12 @@
  * @since 3.4
  */
 @Incubating
+@Deprecated
 public class JavaLanguagePlugin implements Plugin<Project> {
 
     @Override
     public void apply(Project project) {
+        SingleMessageLogger.nagUserOfReplacedPlugin("java-lang", "java-library");
```
Comment (adammurdoch) [inline subprojects/language-java/src/main/java/org/gradle/language/java/plugins/JavaLanguagePlugin.java]: I think we should tweak the message. The `java-library` plugin isn't always the replacement.

I would add a section to the 'upgrading' document to describe what's going on and what to do about it, and then link to that from these deprecation nags and not suggest a specific replacement here.

File: subprojects/plugins/src/testFixtures/groovy/org/gradle/api/tasks/compile/AbstractCachedCompileIntegrationTest.groovy [hunk 3]
```diff
@@ -25,18 +25,24 @@ abstract class AbstractCachedCompileIntegrationTest extends AbstractIntegrationS
         setupProjectInDirectory(temporaryFolder.testDirectory)
     }
 
+    def expectDeprecationWarnings() {
```
Comment (adammurdoch) [inline subprojects/plugins/src/testFixtures/groovy/org/gradle/api/tasks/compile/AbstractCachedCompileIntegrationTest.groovy]: We probably don't need this empty method

**Comment by adammurdoch** [comment]
I think we should document this in some form. We will just need to be careful about the wording.

File: subprojects/plugins/src/testFixtures/groovy/org/gradle/api/tasks/compile/AbstractCachedCompileIntegrationTest.groovy (hunk 3 above)
Comment (jjohannes) [reply subprojects/plugins/src/testFixtures/groovy/org/gradle/api/tasks/compile/AbstractCachedCompileIntegrationTest.groovy]: It's a template method needed for the `CachedPlatformScalaCompileIntegrationTest` subclass.

**Comment by jjohannes** [comment]
Thanks for the feedback @adammurdoch .

I added a section to the upgrading guide: https://github.com/gradle/gradle/pull/11598/commits/9ebe8e4f5d1203b127711e1f3a0d822341fe705a

Before I adjust all tests, I would like to decide on a deprecation warning. Proposal (would add the same message for all deprecations):

```
SingleMessageLogger.nagUserOfDeprecatedPlugin("java-lang", "Consult the upgrading guide for further information: " +
            new DocumentationRegistry().getDocumentationFor("upgrading_version_6", "upgrading_jvm_plugins"));
```

Renders to:

```
The java-lang plugin has been deprecated. This is scheduled to be removed in Gradle 7.0. Consult the upgrading guide for further information: https://docs.gradle.org/6.1/userguide/upgrading_version_6.html#upgrading_jvm_plugins
```

**Pull request merged**